GitUp/Application/AppDelegate.m

```diff
@@ -68,6 +79,113 @@ - (BOOL)canBecomeKeyWindow {
+- (BOOL)canDragItem:(id<NSDraggingInfo>)sender;
+- (NSArray <NSURL *>*)draggingItems:(id<NSDraggingInfo>)sender;
+@end
+
+@implementation GIWelcomeWindow__DraggingInfoExtractor
+- (BOOL)canDragItem:(id<NSDraggingInfo>)sender {
+  return [self draggingItems:sender].count > 0;
+}
+- (NSArray <NSURL *>*)draggingItems:(id<NSDraggingInfo>)sender {
+  return [[sender draggingPasteboard] readObjectsForClasses:@[NSURL.class] options:nil];
+}
+@end
+
+@interface GIWelcomeWindow__DestinationView : NSView
+@property (weak, nonatomic, readwrite) AppDelegate *appDelegate;
+@property (strong, nonatomic, readwrite) GIWelcomeWindow__DraggingInfoExtractor *draggingExtractor;
+@property (assign, nonatomic, readwrite) BOOL receivingDrag;
+@property (strong, nonatomic, readwrite) NSView *imageView;
+@end
+
+@interface GIWelcomeWindow__DestinationView (Configurations)
+- (instancetype)configuredWithAppDelegate:(AppDelegate *)appDelegate;
+@end
+
+@implementation GIWelcomeWindow__DestinationView (Configurations)
+- (instancetype)configuredWithAppDelegate:(AppDelegate *)appDelegate {
+  self.appDelegate = appDelegate;
+  return self;
+}
+@end
+
+@interface GIWelcomeWindow__DestinationView (NSDraggingDestination) <NSDraggingDestination>
+@end
+
+@implementation GIWelcomeWindow__DestinationView
+- (void)setup {
+  self.draggingExtractor = [GIWelcomeWindow__DraggingInfoExtractor new];
+  [self registerForDraggedTypes:@[NSURLPboardType]];
+}
+
+- (void)awakeFromNib {
+  [super awakeFromNib];
+  [self setup];
+}
+
+- (void)setReceivingDrag:(BOOL)receivingDrag {
+  _receivingDrag = receivingDrag;
+  [self setNeedsDisplay:YES];
+}
+
+- (void)drawRect:(NSRect)dirtyRect {
+  if (self.receivingDrag) {
+    [NSColor.selectedControlColor set];
+    NSBezierPath *path = [NSBezierPath bezierPathWithRoundedRect:self.bounds xRadius:10 yRadius:10];
+    path.lineWidth = 5;
+    [path stroke];
+  }
+}
+@end
+
+@implementation GIWelcomeWindow__DestinationView (NSDraggingDestination)
+- (NSDragOperation)draggingEntered:(id<NSDraggingInfo>)sender {
+  BOOL canDragItem = [self.draggingExtractor canDragItem:sender];
+  self.receivingDrag = canDragItem;
+  return canDragItem ? NSDragOperationCopy :  NSDragOperationNone;
+}
+
+- (void)draggingEnded:(id<NSDraggingInfo>)sender {
+  self.receivingDrag = NO;
+}
+
+- (void)draggingExited:(id<NSDraggingInfo>)sender {
+  self.receivingDrag = NO;
+}
+
+- (BOOL)prepareForDragOperation:(id<NSDraggingInfo>)sender {
+  return [self.draggingExtractor canDragItem:sender];
+}
+
+- (BOOL)performDragOperation:(id<NSDraggingInfo>)sender {
+  self.receivingDrag = NO;
+  
+  NSURL *first = [self.draggingExtractor draggingItems:sender].firstObject;
+  [self.appDelegate _openRepositoryWithURL:first withCloneMode:kCloneMode_None windowModeID:NSNotFound];
+  
+  return YES;
+}
+
+- (void)concludeDragOperation:(id<NSDraggingInfo>)sender {
+  // necessary cleanup?
+}
+@end
+
+@interface GIWelcomeWindow : WelcomeWindow
```

Comment:
@lucasderraugh 
Because NSWindow is not intended for drawing. It only contains views.
Do you suggest to draw layer on view in Window?

Reply:
No, I'm saying that `WelcomeWindow` is already an NSWindow subclass that is exclusively used as the welcome window for this app. Why would we create a further subclass? Let's just add the functionality to `WelcomeWindow` instead of `GIWelcomeWindow`